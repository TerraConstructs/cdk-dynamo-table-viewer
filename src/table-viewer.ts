import * as path from "path";
import { Construct } from "constructs";
import {
  Code,
  Runtime,
  EndpointType,
  LambdaRestApi,
  LambdaFunction,
} from "terraconstructs/lib/aws/compute";
import { ITable } from "terraconstructs/lib/aws/storage";

export interface TableViewerProps {
  /**
   * The endpoint type of the
   * [LambdaRestApi](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-apigateway.LambdaRestApi.html)
   * that will be created
   * @default - EDGE
   */
  readonly endpointType?: EndpointType;

  /**
   * The DynamoDB table to view. Note that all contents of this table will be
   * visible to the public.
   */
  readonly table: ITable;

  /**
   * The web page title.
   * @default - No title
   */
  readonly title?: string;

  /**
   * Name of the column to sort by, prefix with "-" for descending order.
   * @default - No sort
   */
  readonly sortBy?: string;

  /**
   * Whether to register the API Gateaway outputs in the stack.
   * @default true
   */
  readonly registerOutputs?: boolean;
}

/**
 * Installs an endpoint in your stack that allows users to view the contents
 * of a DynamoDB table through their browser.
 */
export class TableViewer extends Construct {
  public readonly endpoint: string;

  constructor(parent: Construct, id: string, props: TableViewerProps) {
    super(parent, id);

    const handler = new LambdaFunction(this, "Rendered", {
      code: Code.fromAsset(path.join(__dirname, "..", "lambda")),
      runtime: Runtime.NODEJS_18_X,
      handler: "index.handler",
      environment: {
        TABLE_NAME: props.table.tableName,
        TITLE: props.title || "",
        SORT_BY: props.sortBy || "",
      },
    });

    props.table.grantReadData(handler);

    const home = new LambdaRestApi(this, "ViewerEndpoint", {
      handler,
      endpointConfiguration: props.endpointType
        ? { types: [props.endpointType] }
        : undefined,
      registerOutputs: props.registerOutputs ?? true,
    });
    this.endpoint = home.url;
  }
}
