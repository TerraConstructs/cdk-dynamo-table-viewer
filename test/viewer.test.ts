import * as path from "node:path";
import { ApiGatewayAccount } from "@cdktf/provider-aws/lib/api-gateway-account";
import { ApiGatewayDeployment } from "@cdktf/provider-aws/lib/api-gateway-deployment";
import { ApiGatewayIntegration } from "@cdktf/provider-aws/lib/api-gateway-integration";
import { ApiGatewayMethod } from "@cdktf/provider-aws/lib/api-gateway-method";
import { ApiGatewayResource } from "@cdktf/provider-aws/lib/api-gateway-resource";
import { ApiGatewayRestApi } from "@cdktf/provider-aws/lib/api-gateway-rest-api";
import { ApiGatewayStage } from "@cdktf/provider-aws/lib/api-gateway-stage";
import { DataAwsIamPolicyDocument } from "@cdktf/provider-aws/lib/data-aws-iam-policy-document";
import { DynamodbTable } from "@cdktf/provider-aws/lib/dynamodb-table";
import { IamRole } from "@cdktf/provider-aws/lib/iam-role";
import { LambdaFunction } from "@cdktf/provider-aws/lib/lambda-function";
import { LambdaPermission } from "@cdktf/provider-aws/lib/lambda-permission";
import { App, Testing } from "cdktf";
import "cdktf/lib/testing/adapters/jest";
import { AwsStack } from "terraconstructs/lib/aws/aws-stack";
import * as storage from "terraconstructs/lib/aws/storage";
import * as assertions from "./assertions";
import { TableViewer } from "../src";

const defaultAwsStackProps = {
  environmentName: "test",
  gridUUID: "test-uuid",
  providerConfig: { region: "us-east-1" },
  gridBackendConfig: {
    address: "http://localhost:3000",
  },
};

const TEST_OUTDIR = path.join(__dirname, "cdk.out");
const TEST_APPDIR = path.join(__dirname, "snapshots");
const CDKTFJSON_PATH = path.join(TEST_APPDIR, "cdktf.json");

test("happy  flow", () => {
  // GIVEN
  const app = Testing.stubVersion(
    new App({
      outdir: TEST_OUTDIR,
      stackTraces: false,
      context: {
        cdktfJsonPath: path.resolve(__dirname, CDKTFJSON_PATH),
      },
    }),
  );
  const stack = new AwsStack(app, "test", defaultAwsStackProps);
  const table = new storage.Table(stack, "MyTable", {
    partitionKey: { name: "id", type: storage.AttributeType.STRING },
  });

  // WHEN
  new TableViewer(stack, "viewer", { table });

  // THEN
  // keeping { snapshot: true } so that we can compare changes for v2 upgrade
  const template = new assertions.Template(stack, { snapshot: true });

  template.expect.toHaveResourceWithProperties(LambdaFunction, {
    handler: "index.handler",
    environment: {
      variables: {
        SORT_BY: "",
        TABLE_NAME: "${aws_dynamodb_table.MyTable_794EDED1.name}",
        TITLE: "",
      },
    },
    runtime: "nodejs18.x",
  });
  template.expect.toHaveResourceWithProperties(ApiGatewayRestApi, {
    name: "testviewerViewerEndpointDCCECA7F",
  });
  template.expect.toHaveResourceWithProperties(ApiGatewayStage, {
    stage_name: "prod",
  });
  template.expect.toHaveResourceWithProperties(ApiGatewayResource, {
    path_part: "{proxy+}",
  });

  template.resourceCountIs(ApiGatewayMethod, 2);
  template.expect.toHaveResourceWithProperties(ApiGatewayMethod, {
    http_method: "ANY",
  });
  template.expect.toHaveResourceWithProperties(ApiGatewayIntegration, {
    integration_http_method: "POST",
    type: "AWS_PROXY",
  });
  template.expect.toHaveDataSourceWithProperties(DataAwsIamPolicyDocument, {
    statement: expect.arrayContaining([
      {
        actions: [
          "dynamodb:BatchGetItem",
          "dynamodb:Query",
          "dynamodb:GetItem",
          "dynamodb:Scan",
          "dynamodb:ConditionCheckItem",
          "dynamodb:DescribeTable",
        ],
        effect: "Allow",
        resources: expect.arrayContaining([stack.resolve(table.tableArn)]),
      },
    ]),
  });
  template.resourceCountIs(DynamodbTable, 1);
  template.resourceCountIs(ApiGatewayAccount, 1);
  template.resourceCountIs(ApiGatewayDeployment, 1);
  template.resourceCountIs(IamRole, 2);
  template.resourceCountIs(LambdaPermission, 4);
  const viewerEndpointOutputs = template.outputByName("ViewerEndpointOutputs");
  expect(viewerEndpointOutputs).toBeDefined();
});
