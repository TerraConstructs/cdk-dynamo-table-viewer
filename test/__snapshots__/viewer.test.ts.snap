// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`happy  flow 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_iam_policy_document": {
      "viewer_Rendered_ServiceRole_AssumeRolePolicy_8C6DC3A1": {
        "statement": [
          {
            "actions": [
              "sts:AssumeRole"
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "\${data.aws_service_principal.aws_svcp_default_region_lambda.name}"
                ],
                "type": "Service"
              }
            ]
          }
        ]
      },
      "viewer_Rendered_ServiceRole_DefaultPolicy_7776F322": {
        "statement": [
          {
            "actions": [
              "xray:PutTraceSegments",
              "xray:PutTelemetryRecords"
            ],
            "effect": "Allow",
            "resources": [
              "*"
            ]
          },
          {
            "actions": [
              "dynamodb:BatchGetItem",
              "dynamodb:Query",
              "dynamodb:GetItem",
              "dynamodb:Scan",
              "dynamodb:ConditionCheckItem",
              "dynamodb:DescribeTable"
            ],
            "effect": "Allow",
            "resources": [
              "\${aws_dynamodb_table.MyTable_794EDED1.arn}"
            ]
          }
        ]
      },
      "viewer_ViewerEndpoint_CloudWatchRole_AssumeRolePolicy_4B9372FB": {
        "statement": [
          {
            "actions": [
              "sts:AssumeRole"
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "\${data.aws_service_principal.aws_svcp_default_region_apigateway.name}"
                ],
                "type": "Service"
              }
            ]
          }
        ]
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    },
    "aws_service_principal": {
      "aws_svcp_default_region_apigateway": {
        "service_name": "apigateway"
      },
      "aws_svcp_default_region_lambda": {
        "service_name": "lambda"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "resource": {
    "aws_api_gateway_account": {
      "viewer_ViewerEndpoint_Account_1D965397": {
        "cloudwatch_role_arn": "\${aws_iam_role.viewer_ViewerEndpoint_CloudWatchRole_28045685.arn}",
        "depends_on": [
          "aws_api_gateway_rest_api.viewer_ViewerEndpoint_DED2DEA4"
        ]
      }
    },
    "aws_api_gateway_deployment": {
      "viewer_ViewerEndpoint_Deployment_00B06945": {
        "depends_on": [
          "aws_api_gateway_resource.viewer_ViewerEndpoint_proxy_103C52F5",
          "aws_api_gateway_method.viewer_ViewerEndpoint_ANY_4A9252D2",
          "aws_api_gateway_method.viewer_ViewerEndpoint_proxy_ANY_F7759A2A"
        ],
        "description": "Automatically created by the RestApi construct",
        "lifecycle": {
          "create_before_destroy": true
        },
        "rest_api_id": "\${aws_api_gateway_rest_api.viewer_ViewerEndpoint_DED2DEA4.id}",
        "triggers": {
          "redeployment": "b87a92a4f19666f0bc155c3eda2965af"
        }
      }
    },
    "aws_api_gateway_integration": {
      "viewer_ViewerEndpoint_ANY_Integration_F86AF823": {
        "http_method": "\${aws_api_gateway_method.viewer_ViewerEndpoint_ANY_4A9252D2.http_method}",
        "integration_http_method": "POST",
        "resource_id": "\${aws_api_gateway_rest_api.viewer_ViewerEndpoint_DED2DEA4.root_resource_id}",
        "rest_api_id": "\${aws_api_gateway_rest_api.viewer_ViewerEndpoint_DED2DEA4.id}",
        "type": "AWS_PROXY",
        "uri": "arn:\${data.aws_partition.Partitition.partition}:apigateway:us-east-1:lambda:path/2015-03-31/functions/\${aws_lambda_function.viewer_Rendered_EEFA1E5E.arn}/invocations"
      },
      "viewer_ViewerEndpoint_proxy_ANY_Integration_2F5B634D": {
        "http_method": "\${aws_api_gateway_method.viewer_ViewerEndpoint_proxy_ANY_F7759A2A.http_method}",
        "integration_http_method": "POST",
        "resource_id": "\${aws_api_gateway_resource.viewer_ViewerEndpoint_proxy_103C52F5.id}",
        "rest_api_id": "\${aws_api_gateway_rest_api.viewer_ViewerEndpoint_DED2DEA4.id}",
        "type": "AWS_PROXY",
        "uri": "arn:\${data.aws_partition.Partitition.partition}:apigateway:us-east-1:lambda:path/2015-03-31/functions/\${aws_lambda_function.viewer_Rendered_EEFA1E5E.arn}/invocations"
      }
    },
    "aws_api_gateway_method": {
      "viewer_ViewerEndpoint_ANY_4A9252D2": {
        "authorization": "NONE",
        "http_method": "ANY",
        "resource_id": "\${aws_api_gateway_rest_api.viewer_ViewerEndpoint_DED2DEA4.root_resource_id}",
        "rest_api_id": "\${aws_api_gateway_rest_api.viewer_ViewerEndpoint_DED2DEA4.id}"
      },
      "viewer_ViewerEndpoint_proxy_ANY_F7759A2A": {
        "authorization": "NONE",
        "http_method": "ANY",
        "resource_id": "\${aws_api_gateway_resource.viewer_ViewerEndpoint_proxy_103C52F5.id}",
        "rest_api_id": "\${aws_api_gateway_rest_api.viewer_ViewerEndpoint_DED2DEA4.id}"
      }
    },
    "aws_api_gateway_resource": {
      "viewer_ViewerEndpoint_proxy_103C52F5": {
        "parent_id": "\${aws_api_gateway_rest_api.viewer_ViewerEndpoint_DED2DEA4.root_resource_id}",
        "path_part": "{proxy+}",
        "rest_api_id": "\${aws_api_gateway_rest_api.viewer_ViewerEndpoint_DED2DEA4.id}"
      }
    },
    "aws_api_gateway_rest_api": {
      "viewer_ViewerEndpoint_DED2DEA4": {
        "name": "testviewerViewerEndpointDCCECA7F",
        "tags": {
          "Name": "test-ViewerEndpoint",
          "grid:EnvironmentName": "test",
          "grid:UUID": "test-uuid"
        }
      }
    },
    "aws_api_gateway_stage": {
      "viewer_ViewerEndpoint_DeploymentStageprod_95943DC5": {
        "depends_on": [
          "aws_api_gateway_account.viewer_ViewerEndpoint_Account_1D965397"
        ],
        "deployment_id": "\${aws_api_gateway_deployment.viewer_ViewerEndpoint_Deployment_00B06945.id}",
        "rest_api_id": "\${aws_api_gateway_rest_api.viewer_ViewerEndpoint_DED2DEA4.id}",
        "stage_name": "prod",
        "tags": {
          "Name": "test-ViewerEndpoint",
          "grid:EnvironmentName": "test",
          "grid:UUID": "test-uuid"
        }
      }
    },
    "aws_cloudwatch_log_group": {
      "viewer_Rendered_LogGroup_69DAC943": {
        "name": "/aws/lambda/test-uuid-testviewerRendered",
        "retention_in_days": 7,
        "tags": {
          "Name": "test-Rendered",
          "grid:EnvironmentName": "test",
          "grid:UUID": "test-uuid"
        }
      }
    },
    "aws_dynamodb_table": {
      "MyTable_794EDED1": {
        "attribute": [
          {
            "name": "id",
            "type": "S"
          }
        ],
        "billing_mode": "PROVISIONED",
        "hash_key": "id",
        "name": "testMyTableDB99BBEB",
        "read_capacity": 5,
        "tags": {
          "Name": "test-MyTable",
          "grid:EnvironmentName": "test",
          "grid:UUID": "test-uuid"
        },
        "write_capacity": 5
      }
    },
    "aws_iam_role": {
      "viewer_Rendered_ServiceRole_7AE70B4C": {
        "assume_role_policy": "\${data.aws_iam_policy_document.viewer_Rendered_ServiceRole_AssumeRolePolicy_8C6DC3A1.json}",
        "managed_policy_arns": [
          "arn:\${data.aws_partition.Partitition.partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
        ],
        "name_prefix": "test-uuid-testvieweRenderedServiceRole",
        "tags": {
          "Name": "test-Rendered",
          "grid:EnvironmentName": "test",
          "grid:UUID": "test-uuid"
        }
      },
      "viewer_ViewerEndpoint_CloudWatchRole_28045685": {
        "assume_role_policy": "\${data.aws_iam_policy_document.viewer_ViewerEndpoint_CloudWatchRole_AssumeRolePolicy_4B9372FB.json}",
        "managed_policy_arns": [
          "arn:\${data.aws_partition.Partitition.partition}:iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs"
        ],
        "name_prefix": "test-uuid-testviewepointCloudWatchRole",
        "tags": {
          "Name": "test-ViewerEndpoint",
          "grid:EnvironmentName": "test",
          "grid:UUID": "test-uuid"
        }
      }
    },
    "aws_iam_role_policy": {
      "viewer_Rendered_ServiceRole_DefaultPolicy_ResourceRoles0_39823A44": {
        "name": "testviewerRenderedServiceRoleDefaultPolicyEEF55E0E",
        "policy": "\${data.aws_iam_policy_document.viewer_Rendered_ServiceRole_DefaultPolicy_7776F322.json}",
        "role": "\${aws_iam_role.viewer_Rendered_ServiceRole_7AE70B4C.name}"
      }
    },
    "aws_lambda_function": {
      "viewer_Rendered_EEFA1E5E": {
        "architectures": [
          "x86_64"
        ],
        "depends_on": [
          "aws_cloudwatch_log_group.viewer_Rendered_LogGroup_69DAC943",
          "data.aws_iam_policy_document.viewer_Rendered_ServiceRole_AssumeRolePolicy_8C6DC3A1",
          "aws_iam_role.viewer_Rendered_ServiceRole_7AE70B4C",
          "data.aws_iam_policy_document.viewer_Rendered_ServiceRole_DefaultPolicy_7776F322",
          "aws_iam_role_policy.viewer_Rendered_ServiceRole_DefaultPolicy_ResourceRoles0_39823A44"
        ],
        "environment": {
          "variables": {
            "SORT_BY": "",
            "TABLE_NAME": "\${aws_dynamodb_table.MyTable_794EDED1.name}",
            "TITLE": ""
          }
        },
        "function_name": "test-uuid-testviewerRendered",
        "handler": "index.handler",
        "memory_size": 128,
        "role": "\${aws_iam_role.viewer_Rendered_ServiceRole_7AE70B4C.arn}",
        "runtime": "nodejs18.x",
        "s3_bucket": "\${aws_s3_bucket.AssetBucket.bucket}",
        "s3_key": "\${aws_s3_object.FileAsset_S3.key}",
        "tags": {
          "Name": "test-Rendered",
          "grid:EnvironmentName": "test",
          "grid:UUID": "test-uuid"
        },
        "timeout": 3,
        "tracing_config": {
          "mode": "Active"
        }
      }
    },
    "aws_lambda_permission": {
      "viewer_ViewerEndpoint_ANY_ApiPermissionTesttestviewerViewerEndpointDCCECA7FANY_F9156524": {
        "action": "lambda:InvokeFunction",
        "function_name": "\${aws_lambda_function.viewer_Rendered_EEFA1E5E.arn}",
        "principal": "apigateway.amazonaws.com",
        "source_arn": "arn:\${data.aws_partition.Partitition.partition}:execute-api:us-east-1:\${data.aws_caller_identity.CallerIdentity.account_id}:\${aws_api_gateway_rest_api.viewer_ViewerEndpoint_DED2DEA4.id}/test-invoke-stage/*/"
      },
      "viewer_ViewerEndpoint_ANY_ApiPermissiontestviewerViewerEndpointDCCECA7FANY_731E84A4": {
        "action": "lambda:InvokeFunction",
        "function_name": "\${aws_lambda_function.viewer_Rendered_EEFA1E5E.arn}",
        "principal": "apigateway.amazonaws.com",
        "source_arn": "arn:\${data.aws_partition.Partitition.partition}:execute-api:us-east-1:\${data.aws_caller_identity.CallerIdentity.account_id}:\${aws_api_gateway_rest_api.viewer_ViewerEndpoint_DED2DEA4.id}/\${aws_api_gateway_stage.viewer_ViewerEndpoint_DeploymentStageprod_95943DC5.stage_name}/*/"
      },
      "viewer_ViewerEndpoint_proxy_ANY_ApiPermissionTesttestviewerViewerEndpointDCCECA7FANYproxy_C67CDEE2": {
        "action": "lambda:InvokeFunction",
        "function_name": "\${aws_lambda_function.viewer_Rendered_EEFA1E5E.arn}",
        "principal": "apigateway.amazonaws.com",
        "source_arn": "arn:\${data.aws_partition.Partitition.partition}:execute-api:us-east-1:\${data.aws_caller_identity.CallerIdentity.account_id}:\${aws_api_gateway_rest_api.viewer_ViewerEndpoint_DED2DEA4.id}/test-invoke-stage/*/*"
      },
      "viewer_ViewerEndpoint_proxy_ANY_ApiPermissiontestviewerViewerEndpointDCCECA7FANYproxy_04819798": {
        "action": "lambda:InvokeFunction",
        "function_name": "\${aws_lambda_function.viewer_Rendered_EEFA1E5E.arn}",
        "principal": "apigateway.amazonaws.com",
        "source_arn": "arn:\${data.aws_partition.Partitition.partition}:execute-api:us-east-1:\${data.aws_caller_identity.CallerIdentity.account_id}:\${aws_api_gateway_rest_api.viewer_ViewerEndpoint_DED2DEA4.id}/\${aws_api_gateway_stage.viewer_ViewerEndpoint_DeploymentStageprod_95943DC5.stage_name}/*/*"
      }
    },
    "aws_s3_bucket": {
      "AssetBucket": {
        "bucket": "test-uuid-\${data.aws_caller_identity.CallerIdentity.account_id}-us-east-1"
      }
    },
    "aws_s3_object": {
      "FileAsset_S3": {
        "bucket": "\${aws_s3_bucket.AssetBucket.bucket}",
        "key": "6ef63c8e0feb2ee73d336465b0d50733c8ede8bfdb37d27af4a6582fecd14ed0.zip",
        "source": "assets/FileAsset/6ef63c8e0feb2ee73d336465b0d50733c8ede8bfdb37d27af4a6582fecd14ed0/archive.zip",
        "source_hash": "6ef63c8e0feb2ee73d336465b0d50733c8ede8bfdb37d27af4a6582fecd14ed0"
      }
    }
  },
  "terraform": {
    "backend": {
      "local": {
        "path": "/home/vincent/tcons/cdk-dynamo-table-viewer/terraform.test.tfstate"
      }
    },
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.100.0"
      }
    }
  }
}"
`;
